### ISSUE TITLE
.map on RegEx creates false positive error

### ISSUE THREAD

**Issue by kkmuffme** [microsoft/TypeScript#52825]
# Bug Report

### 🔎 Search Terms

Array map type regex

### 🕗 Version & Regression Information

4.9.5

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/FDBmFcDsGMBcEsD2kAEAjAhgJwBQEoUBvYaZAZ1hVEURQF4UByARgCYBmAQQBYBWANgDsnABwBORgG5gASFKQKKAA4ZYsAKZZUDAPQ4AOgBMA1Hn0B3AybOWjpndJkAbdZSzqy4J5QbVEAOgBbVWgACxxlVQ0tFDxpWXhQFAj3T296OgZILydYolkZd1hwGIBtAF1HAF8QGR0dFHcAc3UADxRg2DD1QxRswLRNMhRkJwBPABoUMlpzdRRoDFR5ADdNSlhafsGsMllUrx9Gj0OgjCUIgDlwAc1Y6RRH4+KYg+9pGqA

### 💻 Code

```js
function bar() {
	const foo = '123A4567A89';
	const pattern = /(\d+)\w(\d+)\w(\d+)/;
	let result = foo.match( pattern );

	if ( result === null ) {
		return [];
	}

	// regex matched numbers only, so we can convert to numbers
	result = result.map( Number );
    return result;
}
```

### 🙁 Actual behavior

Error 2741: Property '0' is missing in type 'number[]' but required in type 'RegExpMatchArray'.

### 🙂 Expected behavior

No error, since the property is there, since .map does not change the array indices.

**Comment by MartinJohns** [comment]
This is not a bug. `RegExpMatchArray` requires to have **at least** one element (the index `0`), but TypeScript has no notation that your `map()` call will return at least one element - it returns an array with arbitrary amount of elements.

**Comment by RyanCavanaugh** [comment]
Probably backing up a bit, TypeScript's type inference rules are that a `let` initialized with some expression puts an upper bound on further assignments to that variable, so just the basic case
```ts
let a = ["foo"];
a = [12];
```
is not allowed by the same rule that creates an error in the given example.
